### PR TITLE
plot of sensor history

### PR DESCRIPTION
**Pull request by silberzwiebel**
This is a first prototype implementation of a plot of sensor history (see #12) I've just learned about the existence of Kotlin with this project and have generally speaking very little experience in Android programming, so feel free to change to the proper coding style whereever needed (and be gentle when reviewing). I've basically copy-pasted and then adapted much of the already existing code without full knowledge ;) 

Also, there are still at least the following major issues (but I wanted to get some overall feedback first), for which I would appreciate any help:

- [x] one cannot select a sensor, it i currently [hardcoded as my temperature sensor](https://opensensemap.org/explore/59c67b5ed67eb50011666dbb). This is because I couldn't figure out how to do this dynamically
- [x] the x- any y-tics labels are not shown in the plot but see [this question at stackoverflow that I posted to resolve this issue](https://stackoverflow.com/questions/47846714/x-and-y-axis-labels-are-not-shown-using-android-widget)
- [x] the x axis should show timestamps, so the API should deliver timestamps and not strings. again, I tried to do this but failed
- [x] the plot is not re-drawn at widget resizing
- [x] overall design and layout of the plot

After these issues are solved, I'd like to actually add some programming logic here, like computing mean values, letting the users decide which timeframe is shown in the plot, showing average values from nearby senseboxes, ...

### CODE REVIEW

**Comment by dbaelz** [comment]
Thanks for the pull request, this is a really great idea. Unfortunately, it will take a few days until I get around to it. I'll review, comment and change the PR as soon as possible.

**Comment by dbaelz** [comment]
Hi, I'm updated some of the code and completed the "re-drawn at widget resizing" item. I could implement the rest or we coordinate the development. Just contact me (here or due the OKFN slack).

**Comment by silberzwiebel** [comment]
Great, thanks for the improvements! If you could implement the proper parsing of the "createdAt" strings as some Time / Date format, I would be happy to use those time values in the plot.
Also, I think it would be easier for both of us, if you could add the full info about the sensebox in the getSensorHistory() call. Then, we could use the information to display proper strings in the plot.

**Comment by dbaelz** [comment]
Current status:
- The `getSenseBoxAndSensorData()` call emits the `SenseBox`, the selected `Sensor` and as List of `SensorHistory` entries for this sensor. This should provide all required data. 
- The models use a DateTime object from the joda time library instead of a String. With this object we can easily format the labels. The [documentation](http://www.joda.org/joda-time/key_format.html) (especially [ISODateTimeFormat](http://www.joda.org/joda-time/apidocs/org/joda/time/format/ISODateTimeFormat.html)) explains the options. When you need some assistants with the format options, then contact me.

**Comment by silberzwiebel** [comment]
Great, thanks! I hope to find time on the weekend to add the missing pieces. Afterwards, I guess, this PR could be merged and I would go on in the next weeks implementing mean values etc. But this might then benefit from individual app settings (e.g., "show last XX hours", "show mean value of day", "show mean value of X closeby senseboxes / near my location").

**Comment by silberzwiebel** [comment]
I've implemented proper strings on the x axis now. Two last smaller issues (but I'm not sure whether they should block the merge):

- It's a bit annoying that the ratio of the widget does not fit the ratio of the plot (which is "thinner"), so there is quite large amounts of black background.
- The x-axis label ("Zeit") is hardcoded and not translatable. I'm not sure on how (and whether) to do this.

**Comment by dbaelz** [comment]
Done
- Fixed a bug concerning empty/null sensor data from the backend.
- Created a string resource (in English) for the x-axis. We should translate the text later on. See #20

Todo
- The plot and the labels are rendered incorrectly on different devices. For example: Due the current `lineLabelInsets.bottom` the x-axis description isn't shown. See screenshot above.
- With the current settings the API-Call for the measurements fetch the last 48 hours and up to 10000 measurements. We should reduce the number of plotted items to increase readability.
- Due the large number of value the plot of the x-axis is more or less useless. See second screenshot above

In my opinion this three todos should be fixed before we merge this pull request. Further improvements could be done in subsequent PRs. I'll try to take care of them at the weekend.

Plot with invisible x-axis description

Plot with x-axis labels

File: app/src/main/java/de/codefor/karlsruhe/opensense/widget/plot/PlotWidget.kt [hunk 1]
```diff
@@ -41,6 +36,11 @@ class PlotWidget : BaseWidget() {
     }
 
     companion object {
+        private val dateTimeFormatter = DateTimeFormat.forPattern("HH:mm")
+        // Use a different formatter without full the date for the first and last tick
+        private val dateTimeFormatterEdges = DateTimeFormat.forPattern("d. M., HH:mm")
```
Comment (silberzwiebel) [inline app/src/main/java/de/codefor/karlsruhe/opensense/widget/plot/PlotWidget.kt]: these format strings are accidentally reversed, the edge formatter should be shorter than the non-edge formatter

**Comment by silberzwiebel** [comment]
> * The plot and the labels are rendered incorrectly on different devices. For example: Due the current lineLabelInsets.bottom the x-axis description isn't shown. See screenshot above.
 > * Due the large number of values the plot of the x-axis is more or less useless. See second screenshot above

Oh, yeah. I've just tested and developed with a tablet (as I don't have a smartphone) and see that it's not looking nice on the phone. Also, the y-axis labels are cut-off.
Although space on a tablet's display is already precious, I imagine it's even more precious on a phone's display.
Given that I don't own such phone, I would be grateful, if you could find nice values for the label placements, I'm happy to check with the tablet afterwards (although arguably a phone is the more likely device for users).
 It would be best to have an automatic solution for this to account for different devices.

> * With the current settings the API-Call for the measurements fetches the last 48 hours and up to 10000 measurements. We should reduce the number of plotted items to increase readability.

I would rather go for a possibility to let the user change these parameters (as I would even like to increase the number of points ;) ). This would require a whole configuration backend, I suppose. What about leaving this parameters now as they are (the default from the sensebox API) and promise a configuration possibility for 0.4?

File: app/src/main/java/de/codefor/karlsruhe/opensense/widget/plot/PlotWidget.kt (hunk 1 above)
Comment (dbaelz) [reply app/src/main/java/de/codefor/karlsruhe/opensense/widget/plot/PlotWidget.kt]: Oh sorry, that was my mistake while refactoring the code. It's fixed with the last commit.

**Comment by dbaelz** [comment]
So I did some improvement and simplification for the plot widget layout. It's not that easy to adjust it for different devices due widget cell size per device. Advanced configuration, dynamical layout etc. are jobs for further opensense versions. In my opinion, the current version is a good start. Please feel free to give some feedback. I'll probably merge the branch at the end of it this week.

**Comment by silberzwiebel** [comment]
Thanks for the improvements and the merging!
I've just find the time now to have a look. Unfortunately, the font is too small now on the tablet and some labels are overlapping (see screenshot below). Couldn't we use the same font size as for the senseBox name on the top / the sensor name on the bottom of the widget? That's a bit too small for the tablet, too, but at least readable.
Also, I'm kind of missing more x-tics.
To save some space: I don't think we need to print the year at the first x-tic, do we?
And last, I actually put the y-axis labels on purpose on the right side, because I thought that's the current value and people might be more interested to quickly judge the accurate value there than the exact value 48 hours ago.

In summary, I guess this all calls for user-definable settings :P 

tmp

**Comment by dbaelz** [comment]
Oh that's a really small font on your device. I tested the design on three devices (Nexus 10, Nexus 7, Nexus 5X) and it looked fine on them. The 10sp from the box name looked a little bit to large, so I reduced it to 7sp. Also, since the font are significantly larger on my devices, I reduce the x-axis label to 3 entries. I rework the font sizes again.

Your thoughts to display y-axis labels on the right side are understandable. I change that, too.
